update/process/shardBlock.go

```diff
@@ -0,0 +1,173 @@
+		return nil, update.ErrNilMarshalizer
+	}
+	if check.IfNil(args.Hasher) {
+		return nil, update.ErrNilHasher
+	}
+	if check.IfNil(args.ShardCoordinator) {
+		return nil, update.ErrNilShardCoordinator
+	}
+	if check.IfNil(args.TxCoordinator) {
+		return nil, update.ErrNilTxCoordinator
+	}
+	if check.IfNil(args.PendingTxProcessor) {
+		return nil, update.ErrNilPendingTxProcessor
+	}
+	if check.IfNil(args.ImportHandler) {
+		return nil, update.ErrNilImportHandler
+	}
+
+	return &shardBlockCreator{
+		shardCoordinator:   args.ShardCoordinator,
+		txCoordinator:      args.TxCoordinator,
+		pendingTxProcessor: args.PendingTxProcessor,
+		importHandler:      args.ImportHandler,
+		marshalizer:        args.Marshalizer,
+		hasher:             args.Hasher,
+	}, nil
+}
+
+// CreateNewBlock will create a new block after hardfork import
+func (s *shardBlockCreator) CreateNewBlock(
+	chainID string,
+	round uint64,
+	nonce uint64,
+	epoch uint32,
+) (data.HeaderHandler, data.BodyHandler, error) {
+	shardHeader := &block.Header{
+		Nonce:           nonce,
+		ShardID:         s.shardCoordinator.SelfId(),
+		Round:           round,
+		Epoch:           epoch,
+		ChainID:         []byte(chainID),
+		AccumulatedFees: big.NewInt(0),
+	}
+
+	blockBody, err := s.createBody()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	shardHeader.MiniBlockHeaders = nil
+	shardHeader.RootHash, err = s.pendingTxProcessor.RootHash()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	shardHeader.ReceiptsHash, err = s.txCoordinator.CreateReceiptsHash()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	totalTxCount, miniBlockHeaders, err := s.createMiniBlockHeaders(blockBody)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	shardHeader.MiniBlockHeaders = miniBlockHeaders
+	shardHeader.TxCount = uint32(totalTxCount)
+
+	metaBlockHash, err := core.CalculateHash(s.marshalizer, s.hasher, s.importHandler.GetHardForkMetaBlock())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	shardHeader.MetaBlockHashes = [][]byte{metaBlockHash}
+
+	return shardHeader, blockBody, nil
+}
+
+func (s *shardBlockCreator) createBody() (*block.Body, error) {
+	mapTxs := s.importHandler.GetTransactions()
+
+	s.txCoordinator.CreateBlockStarted()
+
+	dstMeMiniBlocks, err := s.pendingTxProcessor.ProcessTransactionsDstMe(mapTxs)
+	if err != nil {
+		return nil, err
+	}
+
+	blockBody := &block.Body{
+		MiniBlocks: make([]*block.MiniBlock, 0),
+	}
+
+	postProcessMiniBlocks := s.txCoordinator.CreatePostProcessMiniBlocks()
+	blockBody.MiniBlocks = append(blockBody.MiniBlocks, dstMeMiniBlocks...)
+	blockBody.MiniBlocks = append(blockBody.MiniBlocks, postProcessMiniBlocks...)
+
+	return blockBody, nil
+}
+
+func (s *shardBlockCreator) createMiniBlockHeaders(body *block.Body) (int, []block.MiniBlockHeader, error) {
+	if len(body.MiniBlocks) == 0 {
+		return 0, nil, nil
+	}
+
+	totalTxCount := 0
+	miniBlockHeaders := make([]block.MiniBlockHeader, len(body.MiniBlocks))
+
+	for i := 0; i < len(body.MiniBlocks); i++ {
```

Comment:
So we notarized the the newly generated shard headers inside the first metablock. We might implement this as well for genesis.

Reply:
yes. I thought it is better to save to storage and have all the stuff with it.